File: examples/excel-sheet-metadata.ts
/**
 * 例子（Epic-62）：演示 Sheet 视觉元数据访问器
 * （tabColor / state / activeTab）。
 *
 * 跑法：
 *   bun run examples/excel-sheet-metadata.ts <output.xlsx>
 *
 * 演示内容：
 *   - getSheetState / setSheetState：读写 Sheet 可见性
 *   - getActiveSheet / setActiveSheet：读写活动 sheet 索引
 *   - getWorksheetTabColor / setWorksheetTabColor：读写标签颜色
 *   - clearWorksheetTabColor：清除标签颜色
 */

import { StringValue } from "../src/element/index.js";
import {
  Cell,
  CellValue,
  Row,
  Sheet,
  SheetData,
  Sheets,
  SpreadsheetDocument,
  clearWorksheetTabColor,
  getActiveSheet,
  getSheetState,
  getWorksheetTabColor,
  setActiveSheet,
  setSheetState,
  setWorksheetTabColor,
} from "../src/excel/index.js";

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: excel-sheet-metadata <output.xlsx>\n");
    process.exit(2);
  }

  // 创建工作簿（包含一个默认 Sheet）
  const doc = SpreadsheetDocument.create();
  const workbookPart = doc.workbookPart!;
  const workbook = workbookPart.workbook;

  // 取第一个 worksheet
  const ws = workbookPart.worksheetParts[0]!.worksheet;

  // 写入一行示例数据
  const sheetData = ws.firstChild(SheetData);
  if (sheetData !== undefined) {
    const row = new Row();
    const cell = new Cell();
    cell.cellReference = StringValue.parse("A1");
    cell.dataType = StringValue.parse("str");
    const val = new CellValue();
    val.text = "Epic-62 Sheet 视觉元数据演示";
    cell.appendChild(val);
    row.appendChild(cell);
    sheetData.appendChild(row);
  }

  // 取工作簿中第一个 <x:sheet> 元素
  const sheetsEl = workbook.firstChild(Sheets);
  let sheet: Sheet | undefined;
  if (sheetsEl !== undefined) {
    for (const child of sheetsEl.children) {
      if (child instanceof Sheet) {
        sheet = child;
        break;
      }
    }
  }

  if (sheet === undefined) throw new Error("no Sheet element found");

  // ─── 演示 SheetState ──────────────────────────────────────────────────────
  process.stdout.write(`初始 state: ${getSheetState(sheet)}\n`); // "visible"

  setSheetState(sheet, "hidden");
  process.stdout.write(`set hidden → state: ${getSheetState(sheet)}\n`); // "hidden"

  setSheetState(sheet, "visible");
  process.stdout.write(`set visible → state: ${getSheetState(sheet)}\n`); // "visible"

  // ─── 演示 activeTab ───────────────────────────────────────────────────────
  process.stdout.write(`初始 activeTab: ${getActiveSheet(workbook)}\n`); // 0

  setActiveSheet(workbook, 0);
  process.stdout.write(`set activeSheet(0) → activeTab: ${getActiveSheet(workbook)}\n`); // 0

  // ─── 演示 tabColor ────────────────────────────────────────────────────────
  process.stdout.write(`初始 tabColor: ${String(getWorksheetTabColor(ws))}\n`); // undefined

  setWorksheetTabColor(ws, "FFFF0000"); // 红色
  process.stdout.write(`set FFFF0000 → tabColor: ${getWorksheetTabColor(ws)}\n`); // "FFFF0000"

  setWorksheetTabColor(ws, "FF0000FF"); // 蓝色（覆盖）
  process.stdout.write(`set FF0000FF → tabColor: ${getWorksheetTabColor(ws)}\n`); // "FF0000FF"

  clearWorksheetTabColor(ws);
  process.stdout.write(`clearTabColor → tabColor: ${String(getWorksheetTabColor(ws))}\n`); // undefined

  // ─── 最终状态：红色标签，sheet 保持可见 ──────────────────────────────────
  setWorksheetTabColor(ws, "FFFF0000");
  process.stdout.write(`最终 tabColor: ${getWorksheetTabColor(ws)}\n`);
  process.stdout.write(`最终 state: ${getSheetState(sheet)}\n`);
  process.stdout.write(`最终 activeTab: ${getActiveSheet(workbook)}\n`);

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath}\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/ExcelSheetMetadata.cs
// Replica of examples/excel-sheet-metadata.ts
// Creates xlsx demonstrating tab color set to FFFF0000 (red), sheet visible, activeSheet=0.

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Spreadsheet;

namespace CrossSdkVerify.Replicas;

public static class ExcelSheetMetadata
{
    public static void Run(string outputPath)
    {
        using var doc = SpreadsheetDocument.Create(outputPath,
            DocumentFormat.OpenXml.SpreadsheetDocumentType.Workbook);

        var wbPart = doc.AddWorkbookPart();
        wbPart.Workbook = new Workbook();

        var wsPart = wbPart.AddNewPart<WorksheetPart>();

        // SheetView with workbookViewId=0 (sets activeTab via BookView)
        var sheetViews = new SheetViews(
            new SheetView { TabSelected = true, WorkbookViewId = 0 });

        // Tab color: FFFF0000 (the TS example ends with red after clear+set)
        var sheetPr = new SheetProperties(
            new TabColor { Rgb = "FFFF0000" });

        var sheetData = new SheetData();
        wsPart.Worksheet = new Worksheet();
        wsPart.Worksheet.Append(sheetPr);
        wsPart.Worksheet.Append(sheetViews);
        wsPart.Worksheet.Append(sheetData);

        // Write the single demo cell A1
        var row = new Row { RowIndex = 1 };
        row.Append(new Cell
        {
            CellReference = "A1",
            DataType = CellValues.String,
            CellValue = new CellValue("Epic-62 Sheet 视觉元数据演示")
        });
        sheetData.Append(row);

        var sheets = wbPart.Workbook.AppendChild(new Sheets());
        sheets.Append(new Sheet
        {
            Id = wbPart.GetIdOfPart(wsPart),
            SheetId = 1,
            Name = "Sheet1"
            // State omitted → visible (default)
        });

        // BookView: activeTab=0
        wbPart.Workbook.PrependChild(new BookViews(
            new WorkbookView { ActiveTab = 0 }));

        wbPart.Workbook.Save();
    }
}
